Спринт 13. Функциональность «Рекомендации».

ViktorSalk opened the pull request:
Спринт 13. Функциональность «Рекомендации».

Viktor3642 committed:
Спринт 13. Функциональность «Рекомендации».

1) Создал ветку add-recommendations.
2) Реализовал функции простой рекомендательной системы для фильмов.
3) Реализовал в UserController конечную точку API GET /users/{id}/recommendations.
4) Добавил тесты.

Реализация обеспечивает функциональность для управления рекомендациями, находит пользователей со схожими вкусами, использует простой алгоритм сопоставления, возвращает рекомендации от наиболее похожего пользователя. Постман и прошлые тесты проходит.

polyakovanan commented:
Проверка свалилась, потому что юнит-тесты не прошли)
Не видит RecommendationService в InMemoryFilmControllerTest, InMemoryUserControllerTest и InMemoryReviewControllerTest. Скорее всего, его просто нет с списке для тестового контекста в аннотации SpringBootTest)

Viktor3642 committed:
Спринт 13. Доработал Функциональность «Рекомендации».

1) Обновил InMemoryFilmControllerTest, InMemoryUserControllerTest, InMemoryReviewControllerTest добавив в аннотации SpringBootTest рекомендации.
В IDEA все тесты проходят.

Viktor3642 committed:
Merge remote-tracking branch 'origin/develop' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/film/InMemoryFilmControllerTest.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/review/InMemoryReviewControllerTest.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/user/InMemoryUserControllerTest.java

Viktor3642 committed:
Спринт 13. Доработал Функциональность «Рекомендации».

1) Обновил InMemoryFilmControllerTest, InMemoryUserControllerTest, InMemoryReviewControllerTest добавив в аннотации SpringBootTest рекомендации.
2) Объеденил изменения с веткой лента событий.
В IDEA все тесты проходят.

polyakovanan commented on src/main/java/ru/yandex/practicum/filmorate/service/RecommendationService.java:
```diff
@@ -0,0 +1,86 @@
+                .distinct()
+                .map(filmId -> filmStorage.getById(filmId).orElse(null))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
```
При простом сборе в список лучше вызывать просто toList(), а не collect(). toList() обеспечивает неизменность полученного в ходе выполнения стрима списка.

ViktorSalk replied:
Спасибо)
Заменил  collect() на toList. 
Код стал более лаконичным, а список более защищенным от изменений.

polyakovanan commented on src/test/java/ru/yandex/practicum/filmorate/service/RecommendationServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+
+@ExtendWith(MockitoExtension.class)
+class RecommendationServiceTest {
+    @Mock
```
А мы ведь мокиту не проходили еще)) Оно только в 16 спринте будет)
Но пусть будет, удобная вещь

ViktorSalk replied:
Для нашей задачи искал изолированное тестирование, по соотношению сложности в применении и гибкости, остановился на моке)

polyakovanan reviewed (approved):
Все в полном порядке, только небольшой нюанс с collect().
Можно мержить.

Viktor3642 committed:
Спринт 13. Доработал Функциональность «Рекомендации».

1) Обновил в RecommendationService, collect() на toList,
Код стал более лаконичным, а список более защищенным от изменений.
